src/video/videosurface.cpp

```diff
@@ -68,9 +70,14 @@ void VideoSurface::setSource(VideoSource *src)
     if (source == src)
         return;
 
-    unsubscribe();
-    source = src;
-    subscribe();
+    // To don't block GUI thread.
+    QtConcurrent::run([this, src]()
+    {
+        QMutexLocker l(&bigLock);

```

Comment:
This is wrong. QtConcurrent is designed to be used without external mutexes. Following the code, you will find `bigLock` is relocked, which should cause a race to happen.

Btw. Why do you want to wrap an already running threaded operation in another thread. That totally makes no sense at all. Looks more like something is wrong with the underlying threads, so that the GUI thread is blocked.